core/src/main/scala/kafka/coordinator/GroupCoordinator.scala

```diff
@@ -681,6 +694,17 @@ class GroupCoordinator(val brokerId: Int,
   def onCompleteJoin(group: GroupMetadata) {
     var delayedStore: Option[DelayedStore] = None
     group synchronized {
+      if (group.generationId == 0) {
+        if (group.newMemberAdded && time.milliseconds() < group.initialRebalanceTimeout) {
+          group.newMemberAdded = false
+          val delay = scala.math.min(groupConfig.groupInitialRebalanceDelayMs, group.initialRebalanceTimeout - time.milliseconds()).toInt
+          val groupKey = GroupKey(group.groupId)
+          joinPurgatory.watchOperation(new DelayedJoin(this, group, delay, group.initialRebalanceTimeout), Seq(groupKey))
```

Comment:
Thinking about it a bit more, adding the operation directly into the watch list inside `onCompleteJoin` could result in endless traversal on the list as we are keep re-queueing the operation into the list as we are traversing it, so my previous suggestion does not work perfectly.

I think we need to also change `tryCompleteJoin` to update the `group.initialRebalanceTimeout` and then return false. Such that 1) the calling path from `checkAndComplete` won't reach `onCompleteJoin` and hence would not have endless traversal; 2) the calling path from `ExpiredOperationReaper` will reach `onCompleteJoin` and will add a new operation.

Also I'm wondering why this is not caught in the added `shouldResetRebalanceDelayWhenNewMemberJoinsGroupInInitialRebalance`? Did we actually used a mock purgatory and hence did not expose this issue?